telegram/ext/callbackdatacache.py

```diff
@@ -0,0 +1,423 @@
+                mapped = True
+
+            # Get the cached callback data for the inline keyboard attached to the
+            # CallbackQuery.
+            if callback_query.message:
+                keyboard_uuid = self.__process_message(callback_query.message)
+                for message in (
+                    callback_query.message.pinned_message,
+                    callback_query.message.reply_to_message,
+                ):
+                    if message:
+                        self.__process_message(message)
+                if not mapped and keyboard_uuid:
+                    self._callback_queries[callback_query.id] = keyboard_uuid
+
+    def __get_button_data(self, callback_data: str) -> Any:
+        keyboard, button = self.extract_uuids(callback_data)
+        try:
+            # we get the values before calling update() in case KeyErrors are raised
+            # we don't want to update in that case
+            keyboard_data = self._keyboard_data[keyboard]
+            button_data = keyboard_data.button_data[button]
+            # Update the timestamp for the LRU
+            keyboard_data.update()
+            return button_data
+        except KeyError:
+            return InvalidCallbackData(callback_data)
+
+    def drop_data(self, callback_query: CallbackQuery) -> None:
+        """
+        Deletes the data for the specified callback query.
+
+        Note:
+            Will *not* raise exceptions in case the callback data is not found in the cache.
+            *Will* raise :class:`KeyError` in case the callback query can not be found in the
+            cache.
+
+        Args:
+            callback_query (:class:`telegram.CallbackQuery`): The callback query.
+
+        Raises:
+            KeyError: If the callback query can not be found in the cache
+        """
+        with self.__lock:
+            try:
+                keyboard_uuid = self._callback_queries.pop(callback_query.id)
+                self.__drop_keyboard(keyboard_uuid)
+            except KeyError as exc:
+                raise KeyError('CallbackQuery was not found in cache.') from exc
+
+    def __drop_keyboard(self, keyboard_uuid: str) -> None:
+        try:
+            self._keyboard_data.pop(keyboard_uuid)
+        except KeyError:
+            return
+
+    def clear_callback_data(self, time_cutoff: Union[float, datetime] = None) -> None:
+        """
+        Clears the stored callback data.
+
+        Args:
+            time_cutoff (:obj:`float` | :obj:`datetime.datetime`, optional): Pass a UNIX timestamp
+                or a :obj:`datetime.datetime` to clear only entries which are older.
+                For timezone naive :obj:`datetime.datetime` objects, the default timezone of the
+                bot will be used.
+
+        """
+        with self.__lock:
+            self.__clear(self._keyboard_data, time_cutoff)
+
+    def clear_callback_queries(self) -> None:
```

Comment:
wiki states that this accepts a time_cutoff

Reply:
Maybe my thinking was that it doesn't make as much sense for deleting callback query data … But whatever, to have is better than to need  :D

Reply:
Ah, lol, found the reason: for `_callback_queries` I currently don.t even store the access time. Must have another look, why that is …

Reply:
Okay, I think it's just the fact that CallbackQuery.id is unique and the query will only be handled once in contrast to a button which can be pressed repeatedly. So there is no need to differentiate between "old" and "new" callbackqueries - once you've handled them, you can forget about them.